certificate-authority/service/service.go

```diff
@@ -3,20 +3,70 @@ package service
 import (
 	"context"
 	"fmt"
+	"time"
 
+	gocron "github.com/go-co-op/gocron"
 	grpcService "github.com/plgd-dev/hub/v2/certificate-authority/service/grpc"
 	httpService "github.com/plgd-dev/hub/v2/certificate-authority/service/http"
+	"github.com/plgd-dev/hub/v2/certificate-authority/store/mongodb"
 	"github.com/plgd-dev/hub/v2/pkg/fn"
 	"github.com/plgd-dev/hub/v2/pkg/fsnotify"
 	"github.com/plgd-dev/hub/v2/pkg/log"
 	"github.com/plgd-dev/hub/v2/pkg/net/listener"
 	otelClient "github.com/plgd-dev/hub/v2/pkg/opentelemetry/collector/client"
+	cmClient "github.com/plgd-dev/hub/v2/pkg/security/certManager/client"
 	"github.com/plgd-dev/hub/v2/pkg/security/jwt/validator"
 	"github.com/plgd-dev/hub/v2/pkg/service"
+	"go.opentelemetry.io/otel/trace"
 )
 
 const serviceName = "certificate-authority"
 
+func newStore(ctx context.Context, config StorageConfig, fileWatcher *fsnotify.Watcher, logger log.Logger, tracerProvider trace.TracerProvider) (*mongodb.Store, func(), error) {
+	var fl fn.FuncList
+
+	certManager, err := cmClient.New(config.MongoDB.Mongo.TLS, fileWatcher, logger)
+	if err != nil {
+		return nil, nil, fmt.Errorf("cannot create cert manager: %w", err)
+	}
+	fl.AddFunc(certManager.Close)
+
+	db, err := mongodb.NewStore(ctx, config.MongoDB, certManager.GetTLSConfig(), logger, tracerProvider)
+	if err != nil {
+		fl.Execute()
+		return nil, nil, fmt.Errorf("cannot create mongodb store: %w", err)
+	}
+	fl.AddFunc(func() {
+		if err := db.Close(ctx); err != nil {
+			log.Errorf("failed to close mongodb store: %w", err)
+		}
+	})
+	if config.CleanUpRecords == "" {
+		return db, fl.ToFunction(), nil
+	}
+	s := gocron.NewScheduler(time.Local)
+	if config.ExtendCronParserAboutSeconds {
+		s = s.CronWithSeconds(config.CleanUpRecords)
+	} else {
+		s = s.Cron(config.CleanUpRecords)
+	}
+	_, err = s.Do(func() {
+		_, errDel := db.DeleteNonDeviceExpiredRecords(ctx, time.Now())
+		if errDel != nil {
+			log.Errorf("failed to delete expired signing records: %w", errDel)
+		}
+	})
+	if err != nil {
+		fl.Execute()
+		return nil, nil, fmt.Errorf("cannot create cron job: %w", err)
+	}
+	fl.AddFunc(s.Stop)
+	fl.AddFunc(s.Clear)
```

Comment:
Is this order OK? If I recall the function are executed in reverse order by `fn.FuncList` to mimic deallocation of variables. So `s.Clear` will be called first and `s.Stop`. Maybe `Execute` and `ToFunction` should take a param whether to iterate forwards or backwards.